letsencrypt-auto-source/pieces/pipstrap.py

```diff
@@ -156,12 +156,13 @@ def main():
                                      temp,
                                      digest)
                      for path, digest in PACKAGES]
-        check_output('pip install --no-index --no-deps -U ' +
-                     # Disable cache since we're not using it and it otherwise
-                     # sometimes throws permission warnings:
-                     ('--no-cache-dir ' if has_pip_cache else '') +
-                     ' '.join(quote(d) for d in downloads),
-                     shell=True)
+        # On Windows, pip self-upgrade is not possible, it must be done through python interpreter.
+        command = ['pip'] if name != 'nt' else ['python', '-m', 'pip']
+        command.extend(['install', '--no-index', '--no-deps', '-U'])
+        # Disable cache since it is not used and it otherwise sometimes throws permission warnings:
+        command.extend(['--no-cache-dir'] if has_pip_cache else [])
+        command.extend(downloads)
```

Comment:
I certainly like using the array syntax better than glomming a string together. I like it so much (and always have) that I wonder whether I had a good reason for doing it the other way. Let me check the commit logs.

Reply:
Nope, I did it the worse way from the beginning! Glad you changed it.

Reply:
Your welcome!